src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonDictionaryDefaultConverter.cs

```diff
@@ -0,0 +1,301 @@
+        {
+            JsonConverter<TValue> converter = (JsonConverter<TValue>)state.Current.DeclaredJsonPropertyInfo.ConverterBase;
+            Debug.Assert(converter != null); // It should not be possible to have a null converter at this point.
+
+            return converter;
+        }
+
+        internal override sealed bool OnTryRead(
+            ref Utf8JsonReader reader,
+            Type typeToConvert,
+            JsonSerializerOptions options,
+            ref ReadStack state,
+            out TCollection value)
+        {
+            bool shouldReadPreservedReferences = options.ReferenceHandling.ShouldReadPreservedReferences();
+
+            if (!state.SupportContinuation && !shouldReadPreservedReferences)
+            {
+                // Fast path that avoids maintaining state variables and dealing with preserved references.
+
+                if (reader.TokenType != JsonTokenType.StartObject)
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                }
+
+                CreateCollection(ref state);
+
+                JsonConverter<TValue> elementConverter = GetElementConverter(ref state);
+                if (elementConverter.CanUseDirectReadOrWrite)
+                {
+                    while (true)
+                    {
+                        // Read the key name.
+                        reader.Read();
+
+                        if (reader.TokenType == JsonTokenType.EndObject)
+                        {
+                            break;
+                        }
+
+                        if (reader.TokenType != JsonTokenType.PropertyName)
+                        {
+                            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                        }
+
+                        state.Current.JsonPropertyNameAsString = reader.GetString();
+
+                        // Read the value and add.
+                        reader.Read();
+                        TValue element = elementConverter.Read(ref reader, typeof(TValue), options);
+                        Add(element, options, ref state);
+                    }
+                }
+                else
+                {
+                    while (true)
+                    {
+                        // Read the key name.
+                        reader.Read();
+
+                        if (reader.TokenType == JsonTokenType.EndObject)
+                        {
+                            break;
+                        }
+
+                        if (reader.TokenType != JsonTokenType.PropertyName)
+                        {
+                            ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                        }
+
+                        state.Current.JsonPropertyNameAsString = reader.GetString();
+
+                        // Read the value and add.
+                        reader.Read();
```

Comment:
Yes we had previously discussed this. I tried to split it up into "FastPath" and "SlowPath" methods with [AggressiveInlining] but stepping through the code in release mode, the method was still there are not inlined. Since we now use the stack more often when calling converters (3X stack per converter typically) I left the large method as-is to avoid another stack frame.

Reply:
Sounds good. Let's revisit this later.